manta-pay/src/crypto/poseidon/hash.rs

```diff
@@ -110,18 +158,19 @@ where
     where
         R: Read,
     {
-        Ok(Self::new(
+        Ok(Self::new_unchecked(
```

Comment:
Why using `new_unchecked` instead of `new` here? I don't see there is a check on `ARITY == S::WIDTH + 1` already?

Reply:
Here, we assume that the decoded `permutation` and `domain_tag` are valid.

`new_unchecked(...)` takes two arguments `permutation` and `domain_tag`. Instead, `new(...)` only takes `permutation` and uses `domain_tag(...)` function to generate a `domain_tag` of `ParameterField` type. Since we want to decode both `permutation` and `domain_tag`, we use `new_unchecked`.

Reply:
Yeah but what about the `ARITY` check? I think we still want it here.

Reply:
Yeah but what about the `ARITY` check? I think we still want it here.